src/ChunkingContext.jl

```diff
@@ -0,0 +1,82 @@
+    # byte prefix to skip, used in `skip_rows_init!` and handed to `populate_result_buffer!`
+    # for user to handle with consistently (`_startswith` could be used to do the check)
+    comment::Union{Nothing,Vector{UInt8}}
+    # number of times we refilled the buffer, can be combined with `id` to uniquely identify a chunk
+    buffer_refills::Base.RefValue{Int}
+end
+function ChunkingContext(buffersize::Integer, nworkers::Integer, limit::Integer, comment::Union{Nothing,UInt8,String,Char,Vector{UInt8}})
+    (4 <= buffersize <= typemax(Int32)) || throw(ArgumentError("`buffersize` argument must be larger than 4 and smaller than 2_147_483_648 bytes."))
+    (0 < nworkers < 256) || throw(ArgumentError("`nworkers` argument must be larger than 0 and smaller than 256."))
+    (0 <= limit <= typemax(Int)) || throw(ArgumentError("`limit` argument must be positive and smaller than 9_223_372_036_854_775_808."))
+    # TRACING #  clear_traces!(nworkers)
+    return ChunkingContext(
+        1,
+        TaskCounter(),
+        BufferedVector{Int32}(Int32[0], 1),
+        Vector{UInt8}(undef, buffersize),
+        nworkers,
+        limit,
+        _comment_to_bytes(comment),
+        Ref(0),
+    )
+end
+# Convenience for double-buffering
+function ChunkingContext(ctx::ChunkingContext)
+    out = ChunkingContext(
+        ctx.id + 1,
+        TaskCounter(),
+        BufferedVector{Int32}(Vector{Int32}(undef, max(1, length(ctx.newline_positions))), 1),
+        similar(ctx.bytes),
+        ctx.nworkers,
+        ctx.limit,
+        ctx.comment,
+        Ref(0),
+    )
+    out.newline_positions.elements[1] = 0
+    return out
+end
+tasks_per_chunk(ctx::ChunkingContext) = ctx.nworkers
+total_result_buffers_count(ctx::ChunkingContext) = 2tasks_per_chunk(ctx)
+last_newline_at(ctx::ChunkingContext) = Int(last(ctx.newline_positions))
+function should_use_parallel(ctx::ChunkingContext, _force)
+    return !(
+        _force === :serial ||
+        ((_force !== :parallel) && (Threads.nthreads() == 1 || ctx.nworkers == 1 || last_newline_at(ctx) < MIN_TASK_SIZE_IN_BYTES))
+    )
+end
+
+# We split the detected newlines equally among thr nworkers parsing tasks, but each
+# unit of work should contain at least 16 KiB of raw bytes (MIN_TASK_SIZE_IN_BYTES).
+function estimate_task_size(ctx::ChunkingContext)
+    eols = ctx.newline_positions
+    length(eols) == 1 && return 1 # empty file
+    bytes_to_parse = last(eols)
+    rows = length(eols) # actually rows + 1
+    buffersize = length(ctx.bytes)
+    # There are 2*nworkers result buffers total, but there are nworkers tasks per chunk
+    prorated_maxtasks = ceil(Int, tasks_per_chunk(ctx) * (bytes_to_parse / buffersize))
+    # Lower bound is 2 because length(eols) == 2 => 1 row
+    # bump min rows if average row is much smaller than MIN_TASK_SIZE_IN_BYTES
+    min_rows = max(2, cld(MIN_TASK_SIZE_IN_BYTES, cld(bytes_to_parse, rows)))
+    return min(max(min_rows, cld(rows, prorated_maxtasks)), rows)
```

Comment:
Something still isn't clicking for me here - i'm not sure what this does. Maybe answering my question above would help?

Reply:
I tweaked some of the comments int this file, hopefully its clearer now?